### ISSUE TITLE
Assign File function call under s3 api should add timeout and return error ASAP

### ISSUE THREAD

**Issue by bennykwok** [chrislusf/seaweedfs#2103]
When some of the volume servers are offline (do not meet replication requirement), the native API will return can not assign error immediately.

However, if will get timeout in s3 API. It seem the per assign request wait for 20s (the code is attached below) and it retry for 3 time. As a result I get the the finial assign error after ~60s. 

I think here should get the error immediately and wait for a few second ( may be 5s ? ) and retry for 3 time to ensure that the upload process get the error within 15-30s, so that may do something, such as upload to another backup cluster before client side / gateway timeout.

https://github.com/chrislusf/seaweedfs/blob/c899bdf0636488dc55c6788cc41ce2329f58a39f/weed/server/filer_server_handlers_write_upload.go#L52

for i := 0; i < 3; i++ {
        			// assign one file id for one chunk
        			fileId, urlLocation, auth, assignErr = fs.assignNewFileInfo(so)

**Comment by chrislusf** [comment]
FYI: the 20 seconds comes from here:
https://github.com/chrislusf/seaweedfs/blob/c899bdf0636488dc55c6788cc41ce2329f58a39f/weed/pb/grpc_client_server.go#L71

**Comment by bennykwok** [comment]
yup...I also trace the source of 20 seconds, and added some code to print the assignErr.
It prints 3 time and case a 60s timeout for most client.

But the native API /dir/assign can return error immediately. If s3 api can also do so, the caller function can make decision ASAP.

Moreover, as most of the S3 client / SDK already have a build in retry mechanism. (for example: https://aws.amazon.com/tw/premiumsupport/knowledge-center/lambda-function-retry-timeout-sdk/ and aws cli will retry for 3 time) And so that it may not need to be auto retry in seaweedfs side.

If this action can be controlled by a custom header, such as Aws-maxRetries or Seaweed-maxRetries ( per request timeout may be also controlled by custom header in this way), it may be more flexible.

**Comment by kmlebedev** [comment]
I suggested to put it in the parameters for setting

**Comment by bennykwok** [comment]
if the timeout and retry parameters can be configured by setting file is good. But it's enough.

As this parameters is located in filer , and the s3 api server is on top of filer. And so that the filer may need to do some retry for fuse mount or other task. But it may no need to do retry action for s3.

Besides, I am not sure what will be affect by the 20s timeout, 20s is seem to be too long for some action such as file assign but may be suitable for other action, such as upload or some time consuming action.

Therefore, I suggest chris: 
- the assignErr should be return immediately to filer like what the native API call ( request /dir/assign to a 2/3 volume down system will be error immediately)  to let filer make decision ASAP 
- the filer may keep the current logic in filer to keep backward compatible, but and some logic to let s3 request to bypass / controll the filer build-in retry mechanism and have a shorter and controllable timeout for s3 request
